In [26]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


%matplotlib inline

#### Step 1: Load the dataset

In [27]:
df = pd.read_csv("C:/Users/User/UKM - Nur Azlin Binti Rusnan/Sem 2/Machine Learning/Assignment 2/Assignment2_2023_data.txt", names=[
    'Subject id', 'Jitter (local)', 'Jitter (local, absolute)', 'Jitter (rap)', 'Jitter (ppq5)', 'Jitter (ddp)',
    'Shimmer (local)', 'Shimmer (local, dB)', 'Shimmer (apq3)', 'Shimmer (apq5)', 'Shimmer (apq11)', 'Shimmer (dda)',
    'AC', 'NTH', 'HTN', 'Median pitch', 'Mean pitch', 'Standard deviation', 'Minimum pitch', 'Maximum pitch',
    'Number of pulses', 'Number of periods', 'Mean period', 'Standard deviation of period',
    'Fraction of locally unvoiced frames', 'Number of voice breaks', 'Degree of voice breaks', 'UPDRS score', 'Class Information'
])

In [28]:
df.head(10)

Subject id  Jitter (local)  Jitter (local, absolute)  Jitter (rap)  \
0           1           1.488                  0.000090         0.900   
1           1           0.728                  0.000038         0.353   
2           1           1.220                  0.000074         0.732   
3           1           2.502                  0.000123         1.156   
4           1           3.509                  0.000167         1.715   
5           1           2.470                  0.000126         1.358   
6           1           1.583                  0.000082         0.768   
7           1           1.920                  0.000099         0.926   
8           1           2.257                  0.000119         1.239   
9           1           1.594                  0.000085         0.850   

   Jitter (ppq5)  Jitter (ddp)  Shimmer (local)  Shimmer (local, dB)  \
0          0.794         2.699            8.334                0.779   
1          0.376         1.059            5.864                0.642   
2          0.670         2.196            8.719                0.875   
3          1.634         3.469           13.513                1.273   
4          1.539         5.145            9.112                1.040   
5          1.341         4.075           10.696                1.256   
6          0.864         2.303            9.057                1.070   
7          1.031         2.777           10.184                1.108   
8          1.530         3.716            8.044                0.877   
9          0.951         2.550            7.200                0.823   

   Shimmer (apq3)  Shimmer (apq5)  ...  Maximum pitch  Number of pulses  \
0           4.517           4.609  ...        187.576               160   
1           2.058           3.180  ...        234.505               170   
2           4.347           5.166  ...        211.442              1431   
3           5.263           8.771  ...        220.230                94   
4           3.102           4.927  ...        225.162               117   
5           5.262           7.076  ...        202.812                74   
6           3.794           5.158  ...        200.638                85   
7           3.650           4.611  ...        201.921                71   
8           3.449           4.239  ...        210.523               107   
9           2.480           3.826  ...        203.133               113   

   Number of periods  Mean period  Standard deviation of period  \
0                159     0.006065                      0.000416   
1                169     0.005181                      0.000403   
2               1427     0.006071                      0.000474   
3                 92     0.004910                      0.000320   
4                114     0.004757                      0.000380   
5                 73     0.005118                      0.000187   
6                 84     0.005199                      0.000245   
7                 70     0.005137                      0.000215   
8                104     0.005282                      0.000348   
9                112     0.005335                      0.000332   

   Fraction of locally unvoiced frames  Number of voice breaks  \
0                                0.000                       0   
1                                2.247                       0   
2                               10.656                       1   
3                                0.000                       0   
4                               18.182                       1   
5                               23.214                       0   
6                                4.348                       0   
7                               33.929                       0   
8                               15.152                       2   
9                                0.000                       0   

   Degree of voice breaks  UPDRS score  Class Information  
0                   0.000           23                 

#### Step 2: Explore the Data

In [29]:
# Check for missing values
print("Missing values before handling:\n")
print(df.isnull().sum())

Missing values before handling:

Subject id                             0
Jitter (local)                         0
Jitter (local, absolute)               0
Jitter (rap)                           0
Jitter (ppq5)                          0
Jitter (ddp)                           0
Shimmer (local)                        0
Shimmer (local, dB)                    0
Shimmer (apq3)                         0
Shimmer (apq5)                         0
Shimmer (apq11)                        0
Shimmer (dda)                          0
AC                                     0
NTH                                    0
HTN                                    0
Median pitch                           0
Mean pitch                             0
Standard deviation                     0
Minimum pitch                          0
Maximum pitch                          0
Number of pulses                       0
Number of periods                      0
Mean period                            0
Standard deviation of pe

#### Step 3: Preprocess the Data

In [30]:
# Drop non-significant columns
df = df.drop(['Subject id', 'Class Information'], axis=1)

In [31]:
# Separate features and target variable
X = df.drop(['UPDRS score'], axis=1)
y = df['UPDRS score']

- We dropped 'Subject id', 'UPDRS score', 'Class Information' from the model. 'Subject id' did not significantly impact the MLR model's performance, 'UPDRS score' was designated as the target variable, and 'Class Information' is a categorical variable indicating the presence or absence of Parkinson’s disease, which is not directly relevant for predicting the continuous UPDRS score.

#### Step 4: Split the dataset

In [32]:
from sklearn.model_selection import train_test_split

In [33]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [34]:
from sklearn.preprocessing import StandardScaler

In [35]:
# Standardize the features
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

- We run this step to ensure that all features contribute equally to the model. This is important for the stability and efficiency of the training process in Multiple Linear Regression Model.

#### Step 5: Train the Multiple Linear Regression model

In [36]:
from sklearn.linear_model import LinearRegression

In [37]:
# Build the Multiple Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [38]:
# Predict the UPDRS scores for the test set
y_pred = model.predict(X_test)

#### Step 6: Evaluate the model

In [39]:
from sklearn.metrics import mean_squared_error, r2_score

In [40]:
# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

In [41]:
print(f'Mean Squared Error: {mse}')

Mean Squared Error: 214.1882036951883


MSE measures the average of the squares of the errors that is, the difference between the actual values and the predicted values. In this context, it means that on average, the square of the errors between the predicted and actual UPDRS scores is about 214.19.

A lower MSE indicates better model performance. In this case, an MSE of 214.19 suggests that there is a substantial error in the predictions. This might imply that the model is not performing very well.

In [42]:
print(f'R-squared: {r2}')

R-squared: 0.12217816256786163


R-squared is a statistical measure of how close the data are to the fitted regression line. It represents the proportion of the variance in the dependent variable (UPDRS scores) that is predictable from the independent variables (features).

R² ranges from 0 to 1. An R² value of 0.122 means that only about 12.2% of the variability in UPDRS scores is explained by the model. This indicates that the model is not capturing a significant portion of the variance and might not be a good predictor.

**Model Performance:** 

- The high MSE and low R² values indicate that the model is not performing well. This suggests that either the features are not good predictors or that the linear model is not suitable for this problem.

- Some coefficients are extremely large, indicating potential issues such as multicollinearity or the presence of outliers.

### Steps to Improve the Model

In [6]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import mean_squared_error, r2_score
import statsmodels.api as sm

# Load the dataset

df = pd.read_csv("C:/Users/User/UKM - Nur Azlin Binti Rusnan/Sem 2/Machine Learning/Assignment 2/Assignment2_2023_data.txt", names=[
    'Subject id', 'Jitter (local)', 'Jitter (local, absolute)', 'Jitter (rap)', 'Jitter (ppq5)', 'Jitter (ddp)',
    'Shimmer (local)', 'Shimmer (local, dB)', 'Shimmer (apq3)', 'Shimmer (apq5)', 'Shimmer (apq11)', 'Shimmer (dda)',
    'AC', 'NTH', 'HTN', 'Median pitch', 'Mean pitch', 'Standard deviation', 'Minimum pitch', 'Maximum pitch',
    'Number of pulses', 'Number of periods', 'Mean period', 'Standard deviation of period',
    'Fraction of locally unvoiced frames', 'Number of voice breaks', 'Degree of voice breaks', 'UPDRS score', 'Class Information'
])

# Drop non-significant columns
df = df.drop(['Subject id', 'Class Information'], axis=1)

# Separate features and target variable
X = df.drop(['UPDRS score'], axis=1)
y = df['UPDRS score']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the models
lr = LinearRegression()
ridge = Ridge(alpha=1.0)
lasso = Lasso(alpha=0.1)

# Train and evaluate Linear Regression
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)
mse_lr = mean_squared_error(y_test, y_pred_lr)
r2_lr = r2_score(y_test, y_pred_lr)

# Train and evaluate Ridge Regression
ridge.fit(X_train, y_train)
y_pred_ridge = ridge.predict(X_test)
mse_ridge = mean_squared_error(y_test, y_pred_ridge)
r2_ridge = r2_score(y_test, y_pred_ridge)

# Train and evaluate Lasso Regression
lasso.fit(X_train, y_train)
y_pred_lasso = lasso.predict(X_test)
mse_lasso = mean_squared_error(y_test, y_pred_lasso)
r2_lasso = r2_score(y_test, y_pred_lasso)

# Cross-validation scores
cv_scores_lr = cross_val_score(lr, X, y, cv=5, scoring='r2')
cv_scores_ridge = cross_val_score(ridge, X, y, cv=5, scoring='r2')
cv_scores_lasso = cross_val_score(lasso, X, y, cv=5, scoring='r2')

# Output the results
print("Linear Regression - MSE:", mse_lr, "R2:", r2_lr, "CV R2:", cv_scores_lr.mean())
print("Ridge Regression - MSE:", mse_ridge, "R2:", r2_ridge, "CV R2:", cv_scores_ridge.mean())
print("Lasso Regression - MSE:", mse_lasso, "R2:", r2_lasso, "CV R2:", cv_scores_lasso.mean())

# Checking assumptions with statsmodels
#X_train_sm = sm.add_constant(X_train)
#model = sm.OLS(y_train, X_train_sm).fit()
#print("\n",model.summary())

Linear Regression - MSE: 214.18820369532233 R2: 0.12217816256731229 CV R2: -0.4488644603972037
Ridge Regression - MSE: 218.38497162206056 R2: 0.10497826794581222 CV R2: -0.4090502056005335
Lasso Regression - MSE: 222.1425312800235 R2: 0.08957840994098976 CV R2: -0.4104392544094592


C:\Users\User\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 8.462e+04, tolerance: 2.117e+01
  model = cd_fast.enet_coordinate_descent(
C:\Users\User\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 7.487e+04, tolerance: 1.827e+01
  model = cd_fast.enet_coordinate_descent(
C:\Users\User\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 6.782e+04, tolerance: 

The Mean Squared Error (MSE) values for Linear Regression, Ridge Regression, and Lasso Regression are relatively close to each other, indicating that all three models exhibit similar levels of prediction error. Among these models, Linear Regression has the lowest MSE, suggesting it provides slightly better predictive accuracy compared to Ridge and Lasso Regression for this particular dataset.

The R-squared (R2) values for all models are low, which means that none of the models explain a significant portion of the variance in the dependent variable. Linear Regression has the highest R2 value at 0.122, indicating that it explains only 12.2% of the variance. Ridge and Lasso Regression have even lower R2 values, showing a slight decrease in their explanatory power compared to Linear Regression.

The negative Cross-Validation R2 (CV R2) values for all three models suggest that they do not generalize well to unseen data and may suffer from overfitting. Linear Regression has the lowest (most negative) CV R2 value, indicating that it performs worse in cross-validation compared to Ridge and Lasso Regression.

### **Random Forest**

In [11]:
from sklearn.ensemble import RandomForestRegressor

# Initialize Random Forest Regressor with default parameters
rf_model = RandomForestRegressor(random_state=42)

# Train the model
rf_model.fit(X_train, y_train)

from sklearn.metrics import mean_squared_error

# Predict using the trained model
y_pred_rf = rf_model.predict(X_test)

# Calculate Mean Squared Error
mse_rf = mean_squared_error(y_test, y_pred_rf)

print("Mean Squared Error of Random Forest model:", mse_rf)


Mean Squared Error of Random Forest model: 196.51011875


The Random Forest model with default parameters performed better than the initial MLR model, reducing the MSE from 214.19 to 196.51.

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Load the dataset

data = pd.read_csv("C:/Users/User/UKM - Nur Azlin Binti Rusnan/Sem 2/Machine Learning/Assignment 2/Assignment2_2023_data.txt", names=[
    'Subject id', 'Jitter (local)', 'Jitter (local, absolute)', 'Jitter (rap)', 'Jitter (ppq5)', 'Jitter (ddp)',
    'Shimmer (local)', 'Shimmer (local, dB)', 'Shimmer (apq3)', 'Shimmer (apq5)', 'Shimmer (apq11)', 'Shimmer (dda)',
    'AC', 'NTH', 'HTN', 'Median pitch', 'Mean pitch', 'Standard deviation', 'Minimum pitch', 'Maximum pitch',
    'Number of pulses', 'Number of periods', 'Mean period', 'Standard deviation of period',
    'Fraction of locally unvoiced frames', 'Number of voice breaks', 'Degree of voice breaks', 'UPDRS score', 'Class Information'
])

X = data.drop(columns=['Subject id', 'UPDRS score', 'Class Information'])
y = data['UPDRS score']

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

n_estimators_values = [100, 200, 300]
results = []

for n in n_estimators_values:
    rf_model = RandomForestRegressor(n_estimators=n, random_state=42)
    rf_model.fit(X_train_scaled, y_train)
    y_pred = rf_model.predict(X_test_scaled)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    results.append((n, mse, r2))

best_n = min(results, key=lambda x: x[1])  # Get the parameters with the best (lowest) MSE
print("Best n_estimators:", best_n)


Best n_estimators: (200, 193.3496373798077, 0.20758225232086702)


In [2]:
# Define values for max_depth to test
max_depth_values = [10, 20, 30, None]
results_depth = []

for depth in max_depth_values:
    rf_model = RandomForestRegressor(n_estimators=200, max_depth=depth, random_state=42)
    rf_model.fit(X_train_scaled, y_train)
    y_pred = rf_model.predict(X_test_scaled)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    results_depth.append((depth, mse, r2))

best_depth = min(results_depth, key=lambda x: x[1])  # Get the parameters with the best (lowest) MSE
print("Best max_depth:", best_depth)


Best max_depth: (10, 190.94717165434253, 0.21742843825040026)


In [3]:
# Define values for min_samples_split to test
min_samples_split_values = [2, 5, 10]
results_split = []

for split in min_samples_split_values:
    rf_model = RandomForestRegressor(n_estimators=200, max_depth=10, min_samples_split=split, random_state=42)
    rf_model.fit(X_train_scaled, y_train)
    y_pred = rf_model.predict(X_test_scaled)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    results_split.append((split, mse, r2))

best_split = min(results_split, key=lambda x: x[1])  # Get the parameters with the best (lowest) MSE
print("Best min_samples_split:", best_split)


Best min_samples_split: (2, 190.94717165434253, 0.21742843825040026)


In [4]:
# Define values for min_samples_leaf to test
min_samples_leaf_values = [1, 2, 4]
results_leaf = []

for leaf in min_samples_leaf_values:
    rf_model = RandomForestRegressor(n_estimators=200, max_depth=10, min_samples_split=2, min_samples_leaf=leaf, random_state=42)
    rf_model.fit(X_train_scaled, y_train)
    y_pred = rf_model.predict(X_test_scaled)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    results_leaf.append((leaf, mse, r2))

best_leaf = min(results_leaf, key=lambda x: x[1])  # Get the parameters with the best (lowest) MSE
print("Best min_samples_leaf:", best_leaf)


Best min_samples_leaf: (4, 186.62558429747136, 0.23513988868877633)


In [5]:
# Train the final model with the best hyperparameters
final_rf_model = RandomForestRegressor(
    n_estimators=200,
    max_depth=10,
    min_samples_split=2,
    min_samples_leaf=4,
    random_state=42
)
final_rf_model.fit(X_train_scaled, y_train)

# Predict on the test set
final_y_pred = final_rf_model.predict(X_test_scaled)

# Evaluate the final model
final_mse = mean_squared_error(y_test, final_y_pred)
final_r2 = r2_score(y_test, final_y_pred)

final_mse, final_r2


(186.62558429747136, 0.23513988868877633)

**Model Performance:** 

**Multiple Linear Regression (MLR) Model:**

The MLR model achieved an MSE of approximately 214.19 on the test set. While it provides a straightforward and interpretable approach, its performance in terms of MSE was not optimal compared to other models.

**Random Forest Model:**

Using default parameters, the Random Forest model reduced the MSE to 196.51. Further tuning of the model improved the MSE to 186.63. The final Random Forest model achieved an R² score of 0.235, indicating that it explained about 23.5% of the variance in UPDRS scores.

**Conclusion:**

Exploring various models revealed that the Random Forest model significantly outperformed the MLR model, which had an MSE of 214.19. This comparison highlights the potential limitations of the MLR model for this dataset and suggests that more complex models, like Random Forests, may offer better predictive performance.